doc/configuration.md
# botvinnik: configuration file

The configuration file of botvinnik is the place where the settings like
the Matrix homeserver or the user id and password for login are stored.
**Since this is sensitive information, take precaution that no other
users or only users who you trust can access the configuration file.** This can
usually be achieved by setting proper file permissions and file ownership for
the configuration file. (Something along the lines of `chmod 600 FILE_NAME` is
the usual approach, together with `chown user:group FILE_NAME`.)

# Location of the configuration file

The location of the configuration file can be specified via the command line
option `-c`. For example

    /path/to/botvinnik -c /home/user/some/path/bvn.conf

will tell the program to look for the configuration file
`/home/user/some/path/bvn.conf`. If no such option is given, the program will
look for the configuration file in these predefined locations, in the given
order:

* `user's home directory`/.bvn/bvn.conf
* `user's home directory`/.botvinnik/bvn.conf
* `user's home directory`/.botvinnik/botvinnik.conf
* /etc/botvinnik/botvinnik.conf
* /etc/botvinnik/bvn.conf
* /etc/bvn/bvn.conf
* ./botvinnik.conf
* ./bvn.conf

If none of these files exist and no location is given for the configuration
file, then the program will exit with a non-zero exit code, because a
configuration file is required for the bot to work properly.

# Configuration elements

## General format

Configuration settings are basically key-value pairs, where the key is specified
first, followed by an equals sign (`=`), followed by the value of the setting.
For example, the line

    matrix.homeserver=https://matrix.example.org/

sets the value of `matrix.homeserver` to `https://matrix.example.org/`.

## Comments

The configuration file allows comment lines. Comment lines are lines where the
first character is `#`. These lines are ignored by the application. A line has
to be either a comment or configuration data, you cannot mix both on the same
line.

Completely empty lines are ignored, too.

## Matrix server login settings

The following Matrix-related settings are recognized in the configuration file:

* **matrix.homeserver** - _(since 0.0.1, required)_ host name of the server
  where the Matrix instance runs (e. g. `https://matrix.example.tld/`)
* **matrix.userid** - _(since 0.0.1, required)_ user id of the Matrix user
  (e. g. `@alice:example.tld`)
* **matrix.password** - _(since 0.0.1, required)_ password for the Matrix user

All of these three settings are required to be present in any valid
configuration file, because these are essential for connecting to the Matrix
homeserver. The bot will post its replies as the Matrix user specified by the
**matrix.userid** setting. It is therefore advisable (although not required) to
create a separate Matrix user that is just used by the bot.

## Bot management settings

* **command.prefix** - _(since 0.0.5, optional)_ prefix for text commands (e. g.
  if you have a hypothetical command `foo` and the prefix is set to `!`, then
  the text message `!foo` will trigger that command); defaults to `!` if not set
* **command.deactivate** - _(since 0.5.0, optional)_ specifies a bot command
  that will be deactivated, i. e. it will not be available when the bot is
  running. For example, the line `command.deactivate=ping` would deactivate the
  `!ping` command. This option can be given multiple times to deactivate more
  than one command.

  _Note:_ Some commands that are essential to operate the bot cannot be
  deactivated. The bot will not start if you attempt to deactivate such a
  command. Those commands are: `!help`, `!leave`, `!rooms`, `!stop`, `!version`,
  and `!whoami`.
* **bot.stop.allowed.userid** - _(since 0.0.11, required)_ user id of a Matrix
  user that is allowed to stop the bot (e. g. `@alice:example.tld`). This
  setting may occur multiple times with different user ids to allow more users
  to stop the bot. At least one user id must be specified. Furthermore, the user
  id from the setting **matrix.userid** will be added to this list
  automatically, if it has not been specified as user allowed to stop the bot.
* **bot.sync.allowed_failures** - _(since 0.2.0, optional)_ the number of
  allowed Matrix synchronization failures within the last 64 synchronization
  attempts before the bot exits. Allowed range is from 0 (zero) to 63. This
  setting can be used to avoid that the bot stops when the network connection
  may occasionally be down for a few seconds. One allowed failure is approx. the
  equivalent of 2.5 seconds network downtime. (This time may change, if the
  setting **bot.sync.delay_milliseconds** is set to something else than 2500.)
  Default value of allowed failures is 24, if it is not set explicitly.

  Set this to zero, if you want to restore the behaviour used in versions before
  0.2.0. However, a value of zero means that the bot stops whenever there is a
  small, short network downtime, so this is not advisable.
* **bot.sync.delay_milliseconds** - _(since 0.5.0, optional)_ the delay between
  two consecutive Matrix synchronization requests in milliseconds. (One second
  consists of 1000 milliseconds, e. g. 5000 milliseconds are 5 seconds.)
  Allowed range is from 100 (0.1 seconds) to 30000 (30 seconds). Values outside
  of this range will be replaced by the allowed minimum or maximum - whatever is
  closer. Default value is 2500, if it is not set explicitly.

  Primary purpose of this setting is to allow a balance between too high load on
  the Matrix server due to too many synchronization requests from the bot on the
  one side and a bot that is responding to commands too slowly on the other
  side.

## Translation server settings

The following settings control how the text translation commands work:

* **libretranslate.server** - _(since 0.4.0, optional)_ URL of the
  LibreTranslate server to use for translation commands. If this is omitted,
  then it is assumed to be empty. Empty server URL means no translation will be
  available.
* **libretranslate.apikey** - _(since 0.4.0, optional)_ API key for the
  LibreTranslate server. If this setting is omitted, then it is assumed to be
  empty. If this is empty, then no API key will be used in requests to the
  translation server. Some LibreTranslate instances will not work without an API
  key.

Note that at least a translation server URL has to be set, if you want the users
of the bot to be able to use the translation feature. Some servers also require
an API key, while other servers work without one.

A list of some possible LibreTranslate servers is available at
<https://github.com/LibreTranslate/LibreTranslate#mirrors>.

## Giphy settings

To use the Giphy plugin, the bot needs a valid API key for Giphy. Without it,
the plugin will not work.

* **giphy.apikey** - _(since 0.6.3, optional)_ API key for Giphy. If this
  setting is omitted, then it is assumed to be empty. If this is empty, then no
  requests to the Giphy API will be performed. The plugin will show a message
  about a missing API key instead when it's command `!giphy` is invoked.

# Example of a complete configuration file

The following example is a complete configuration file for the botvinnik program
(as of version 0.6.3 or later):

    # This line is a comment and will be ignored by the program.
    #And so is this line.

    # Matrix server login settings
    matrix.homeserver=https://matrix.example.tld/
    matrix.userid=@alice:matrix.example.tld
    matrix.password=secret, secret, top(!) secret
    # bot management settings
    command.prefix=!
    bot.stop.allowed.userid=@bob:matrix.example.tld
    bot.sync.allowed_failures=12
    bot.sync.delay_milliseconds=5000
    # deactivate the !ping command
    command.deactivate=ping
    # translation server settings
    libretranslate.server=https://libretranslate.com
    libretranslate.apikey=abcdef1234567890
    # Giphy API key
    giphy.apikey=AbCdEfGhIjKlMnOpQrStUvWxYz123456

# Example of a minimal configuration file

The following example is a minimal configuration file for the botvinnik program
(as of version 0.6.3 or later). It only contains the configuration settings
which are required to start the bot. Any optional settings are left out.

    # Matrix server login settings
    matrix.homeserver=https://matrix.example.tld/
    matrix.userid=@alice:matrix.example.tld
    matrix.password=secret, secret, top(!) secret
    # bot management settings
    bot.stop.allowed.userid=@bob:matrix.example.tld


changelog.md
# Version history of botvinnik Matrix bot

_(Note: This changelog focuses on the major changes between the different
versions. Therefore, it may not contain all changes. Especially smaller fixes or
improvements may be omitted.)_

## Version 0.9.2 (2024-03-12)

* __[feature]__
  Two more languages are now supported by the Wikipedia command: Japanese and
  Chinese. For those the new commands `!wikija` (for the Japanese Wikipedia)
  and `!wikizh` (for the Chinese Wikipedia) are added.

* __[maintenance]__
  The library that does the JSON parsing (simdjson) has been updated from
  version 3.7.0 to version 3.8.0.

## Version 0.9.1 (2024-03-09)

* __[fix]__
  Fix a bug where some Linux distributions (namely Alpine) could not properly
  detect the user's home directory and thus would not search for an existing
  configuration file in the user's home directory.

## Version 0.9.0 (2024-03-06)

* __[fix]__
  The `!fortune` command will now work on Linux distributions that are not based
  on Debian, too.

* __[change]__
  The `!fortune` command is now only available, if the bot runs on a Linux
  system. Technically, that is not a limitation, because the fortune binary is
  usually not available on Windows or macOS systems.

## Version 0.8.0 (2024-03-01)

* __[change]__
  The data source for the `!corona` command has been switched from the World
  Health Organization to Our World in Data. Reason for that switch is that the
  World Health Organization stopped providing daily data and only provides
  weekly data now.

  As part of that change, the option to specify a country by its two-letter
  country code according to ISO-3166 is changed to three-letter code instead.
  That's used in the data which is provided by Our World in Data, so the bot
  will use that from now on. The change means that e. g. when users previously
  used the command `!corona GR` to get the numbers for Greece, they will now
  have to use `!corona GRC` instead. Of course, full country names are still
  possible, too, so `!corona Greece` will also work in that case.

* __[maintenance]__
  The library that does the JSON parsing (simdjson) has been updated from
  version 3.6.3 to version 3.7.0.

## Version 0.7.0 (2023-12-10)

* __[new feature]__
  The `!help` command can now also show a detailed help for a specific command,
  if a command is named. For example, the command `!help wiki` or `!help !wiki`
  will show detailed help for the `!wiki` command.

  If the the `!help` command is sent without a command name it will list all
  available commands and a short(er) explanation for each command, just as in
  previous versions of the bot.

* __[improvement]__
  The images that the bot sends as response to the `!giphy` command will now
  have some metadata about the image size in bytes, the width and the height.

* __[maintenance]__
  The library that does the JSON parsing (simdjson) has been updated from
  version 3.2.3 to version 3.6.3.

* __[maintenance]__
  The library that does the JSON serialization (nlohmann/json) has been updated
  from version 3.11.2 to version 3.11.3.

## Version 0.6.3 (2023-09-01)

* __[new feature]__
  A new command to display random GIFs - `!giphy` - is added. For example, the
  command `!giphy cat` will show a random GIF of a cat.

* __[maintenance]__
  The library that does the JSON parsing (simdjson) has been updated from
  version 3.2.0 to version 3.2.3.

## Version 0.6.2 (2023-06-25)

* __[new feature]__
The Debian package search is extended by adding the `!deb13` command to search
for packages in Debian 13 (a.k.a. "trixie").

* __[change]__
Furthermore, the `!deb` command is now an alias for `!deb12` instead for
`!deb11`. In other words: `!deb` does now search packages of Debian 12 instead
of Debian 11 as before.

* __[improvement]__
  When requesting COVID-19 data for Kosovo via the `!corona` command, the
  country will now be correctly named "Kosovo" instead of "Kosovo[1]". This is a
  quirk in the WHO data, and the program will now work around it.

* __[maintenance]__
  The library that does the JSON parsing (simdjson) has been updated from
  version 3.1.7 to version 3.2.0.

* __[maintenance]__
  The base image for the Docker build is updated from `debian:11-slim` to
  `debian:12-slim`, i. e. from Debian 11 (codename "bullseye") to Debian 12
  (codename "bookworm").

## Version 0.6.1 (2023-05-03)

__[improvement]__

Remove superfluous / redundant data from database of COVID-19 data created by
the `!corona` command to speed up database creation.

Furthermore, the search algorithm for finding countries by name has been
improved to allow for some variants of longer, complex names.

## Version 0.6.0 (2023-04-30)

* __[change]__
  The data source for the `!corona` command has been switched from the Center
  for Systems Science and Engineering (CSSE) at Johns Hopkins University to the
  World Health Organization COVID-19 Dashboard.
  Reason for that switch is that Johns Hopkins University stopped the collection
  of new data on 10th March 2023. The WHO also provides data beyond that point.

  The change to WHO as data source affects a few countries / territories
  (two-letter country codes according to ISO-3166 are given in parenthesis for
   easier search / identification):

  * Bonaire, Saint Eustatius and Saba (BQ) is split into three parts which now
    can be queried individually:
    * Bonaire (XA)
    * Sint Eustatius (XB)
    * Saba (XC)
  * Cape Verde (CV) is now called Cabo Verde.
  * Cote d'Ivoire (CI) is now called Côte d’Ivoire.
  * French Guiana (GF) is added as new country / territory.
  * Guadeloupe (GP) is added as new country / territory.
  * Laos (LA) is now called Lao People's Democratic Republic.
  * Martinique (MQ) is added as new country / territory.
  * Mayotte (YT) is added as new country / territory.
  * Micronesia (FM) is now called Micronesia (Federated States of).
  * Nauru (NR) is added as new country / territory.
  * Niue (NU) is added as new country / territory.

src/conf/Configuration.cpp
/*
 -------------------------------------------------------------------------------
    This file is part of the botvinnik Matrix bot.
    Copyright (C) 2020, 2022, 2023, 2024  Dirk Stolle

    This program is free software: you can redistribute it and/or modify
    it under the terms of the GNU General Public License as published by
    the Free Software Foundation, either version 3 of the License, or
    (at your option) any later version.

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program.  If not, see <http://www.gnu.org/licenses/>.
 -------------------------------------------------------------------------------
*/

#include "Configuration.hpp"
#include <filesystem>
#include <fstream>
#include <iostream>
#include "../util/Directories.hpp"
#include "../util/Strings.hpp"

namespace bvn
{

bool looksLikeUserId(const std::string& str)
{
  // User id must start with "@".

  // std::string::at(0) will not throw, because string is checked for emptiness
  // before it lands here.
  if (str.at(0) != '@')
    return false;
  // User id must contain a ':'.
  return str.find(':') != std::string::npos;
}

// use same comment character as in task files: '#'
const char Configuration::commentCharacter = '#';

// Default value for allowed failures.
const int Configuration::default_allowed_failures = 24;

// Anything below 100 ms is awfully close to a DoS attempt.
const std::chrono::milliseconds Configuration::min_sync_delay = std::chrono::milliseconds(100);

// Default synchronization delay.
const std::chrono::milliseconds Configuration::default_sync_delay = std::chrono::milliseconds(2500);

// Anything above 30 seconds (=30000 ms) is too slow to be recognized as a
// timely response.
const std::chrono::milliseconds Configuration::max_sync_delay = std::chrono::milliseconds(30000);

Configuration::Configuration()
:
  mHomeServer(""),
  mUserId(""),
  mPassword(""),
  mPrefix(""),
  mDeactivatedCommands(std::unordered_set<std::string>()),
  mStopUsers(std::unordered_set<std::string>()),
  mAllowedFailsIn64(-1),
  mSyncDelay(std::chrono::milliseconds::zero()),
  mLibreTranslateServer(""),
  mLibreTranslateApiKey(""),
  mGiphyApiKey("")
{
}

std::vector<std::string> Configuration::potentialFileNames()
{
  std::vector<std::string> result = {
      "/etc/botvinnik/botvinnik.conf",
      "/etc/botvinnik/bvn.conf",
      "/etc/bvn/bvn.conf",
      "./botvinnik.conf",
      "./bvn.conf"
  };
  std::string home;
  if (filesystem::getHome(home))
  {
    const auto delim = filesystem::pathDelimiter;
    result.insert(result.begin(), home + delim + ".botvinnik" + delim + "botvinnik.conf");
    result.insert(result.begin(), home + delim + ".botvinnik" + delim + "bvn.conf");
    result.insert(result.begin(), home + delim + ".bvn" + delim + "bvn.conf");
  }
  return result;
}

const std::string& Configuration::homeServer() const
{
  return mHomeServer;
}

const std::string& Configuration::userId() const
{
  return mUserId;
}

const std::string& Configuration::password() const
{
  return mPassword;
}

const std::string& Configuration::prefix() const
{
  return mPrefix;
}

const std::unordered_set<std::string>& Configuration::deactivatedCommands() const
{
  return mDeactivatedCommands;
}

const std::unordered_set<std::string>& Configuration::stopUsers() const
{
  return mStopUsers;
}

bool Configuration::isAdminUser(const std::string& userId) const
{
  return mStopUsers.find(userId) != mStopUsers.end();
}

int Configuration::allowedFailures() const
{
  return mAllowedFailsIn64;
}

std::chrono::milliseconds Configuration::syncDelay() const
{
  return mSyncDelay;
}

const std::string& Configuration::translationServer() const
{
  return mLibreTranslateServer;
}

const std::string& Configuration::translationApiKey() const
{
  return mLibreTranslateApiKey;
}

const std::string& Configuration::gifApiKey() const
{
  return mGiphyApiKey;
}

void Configuration::findConfigurationFile(std::string& realName)
{
  namespace fs = std::filesystem;

  if (realName.empty())
  {
    for(const auto& file : potentialFileNames())
    {
      const fs::path path(file);
      try
      {
        if (fs::exists(path) && fs::is_regular_file(path))
        {
          realName = file;
          std::cout << "Info: Using configuration file " << file
                    << ", because none has been specified explicitly." << std::endl;
          break;
        }
      }
      catch (const fs::filesystem_error& ex)
      {
        std::cerr << "File system error while checking existence of file "
                  << file << ": " << ex.what() << std::endl;
      } // try-catch
    } // for
  } // if
}

bool Configuration::loadCoreConfiguration(const std::string& fileName)
{
  std::ifstream stream(fileName, std::ios_base::in | std::ios_base::binary);
  if (!stream.good())
  {
    std::cerr << "Error: Could not open configuration file " << fileName << "!" << std::endl;
    return false;
  }

  std::string line;
  while (std::getline(stream, line))
  {
    trim(line);
    // skip empty lines and comment lines
    if (line.empty() || line[0] == commentCharacter)
      continue;

    // check for possible carriage return at end (happens on Windows systems)
    if (line.at(line.length() - 1) == '\r')
    {
      line.erase(line.length() - 1);
    } // if

    const auto sepPos = line.find('=');
    if (sepPos == std::string::npos)
    {
      std::cerr << "Error: Invalid line found: \"" << line <<"\".\n"
                << "General format: \"Name of Setting=value\"" << std::endl;
      return false;
    }
    std::string name = line.substr(0, sepPos);
    trim(name);
    std::string value = line.substr(sepPos);
    value.erase(0, 1);
    trim(value);

    if (value.empty())
    {
      std::cerr << "Error: Value for " << name << " must not be empty!" << std::endl;
      return false;
    }

    if ((name == "matrix.homeserver") || (name == "homeserver"))
    {
      if (!mHomeServer.empty())
      {
        std::cerr << "Error: Matrix homeserver is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      // Remove trailing slash.
      if (value.size() > 0 && value.at(value.size() - 1) == '/')
        value.erase(value.size() - 1);

      if (value.substr(0, 7) == "http://")
      {
        std::clog << "Warning: Homeserver URL does not use HTTPS!" << std::endl;
      }
      else if (value.substr(0, 8) != "https://")
      {
        value = "https://" + value;
        std::cout << "Info: Prepending 'https://' to homeserver URL, it's now "
                  << value <<"." << std::endl;
      }
      mHomeServer = value;
    } // if matrix.homeserver
    else if ((name == "matrix.userid") || (name == "userid"))
    {
      if (!mUserId.empty())
      {
        std::cerr << "Error: Matrix user id is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      if (!looksLikeUserId(value))
      {
        std::cerr << "Error: Configuration value of " << name
                  << " does not seem to be a Matrix user id!" << std::endl;
        return false;
      }
      mUserId = value;
    } // if matrix.userid
    else if ((name == "matrix.password") || (name == "password"))
    {
      if (!mPassword.empty())
      {
        std::cerr << "Error: Password is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      mPassword = value;
    } // if matrix.password
    else if ((name == "command.prefix") || (name == "prefix"))
    {
      if (!mPrefix.empty())
      {
        std::cerr << "Error: Command prefix is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      mPrefix = value;
    } // if command.prefix
    else if ((name == "command.deactivate") || (name == "command.deactivated"))
    {
      if (mDeactivatedCommands.find(value) != mDeactivatedCommands.end())
      {
        std::cerr << "Error: The command '" << value << "' is deactivated more"
                  << " than once in file " << fileName << "!" << std::endl;
        return false;
      }
      // Avoid adding non-existent commands over and over again.
      if (mDeactivatedCommands.size() >= 100)
      {
        std::cerr << "Error: There are too many deactivated commands in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      mDeactivatedCommands.insert(value);
    } // if command.deactivate
    else if ((name == "bot.stop.allowed.userid") || (name == "stop.allowed.userid"))
    {
      if (mStopUsers.find(value) != mStopUsers.end())
      {
        std::cerr << "Error: Matrix user id '" << value << "' for user that is"
                  << " allowed to stop the bot is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      if (!looksLikeUserId(value))
      {
        std::cerr << "Error: Configuration value of " << name
                  << " does not seem to be a Matrix user id!" << std::endl;
        return false;
      }
      mStopUsers.insert(value);
    } // if bot.stop.allowed.userid
    else if ((name == "bot.sync.allowed_failures") || (name == "bot.allowed_failures"))
    {
      if (mAllowedFailsIn64 >= 0)
      {
        std::cerr << "Error: Number of allowed sync failures is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      if (!stringToInt(value, mAllowedFailsIn64))
      {
        std::cerr << "Error: Number of allowed sync failures in file "
                  << fileName << " must be a non-negative integer!" << std::endl;
        return false;
      }
      // If it is still below zero or above 63, the value is invalid.
      if (mAllowedFailsIn64 < 0 || mAllowedFailsIn64 > 63)
      {
        std::cerr << "Error: Number of allowed sync failures in file "
                  << fileName << " must be between 0 and 63 (inclusive)!" << std::endl;
        return false;
      }
    } // if bot.sync.allowed_failures
    else if ((name == "bot.sync.delay_milliseconds") || (name == "bot.delay_milliseconds"))
    {
      if (mSyncDelay != std::chrono::milliseconds::zero())
      {
        std::cerr << "Error: Synchronization delay is specified more than once"
                  << " in file " << fileName << "!" << std::endl;
        return false;
      }
      int amount_of_ms = 0;
      if (!stringToInt(value, amount_of_ms))
      {
        std::cerr << "Error: Synchronization delay in file " << fileName
                  << " must be a non-negative integer!" << std::endl;
        return false;
      }
      // If it is outside of the allowed minimum or maximum, the value is
      // invalid and will be clamped.
      if (amount_of_ms < min_sync_delay.count())
      {
        std::clog << "Warning: Synchronization delay in file " << fileName
                  << " is lower than allowed and will be raised to the allowed"
                  << " minimum of " << min_sync_delay.count()
                  << " milliseconds!" << std::endl;
        amount_of_ms = min_sync_delay.count();
      }
      else if (amount_of_ms > max_sync_delay.count())
      {
        std::clog << "Warning: Synchronization delay in file " << fileName
                  << " is higher than allowed and will be lowered to the"
                  << " allowed maximum of " << max_sync_delay.count()
                  << " milliseconds!" << std::endl;
        amount_of_ms = max_sync_delay.count();
      }
      mSyncDelay = std::chrono::milliseconds(amount_of_ms);
    } // if bot.sync.delay_milliseconds
    else if (name == "libretranslate.server")
    {
      if (!mLibreTranslateServer.empty())
      {
        std::cerr << "Error: Server URL for LibreTranslate is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      mLibreTranslateServer = value;
    } // if libretranslate.server
    else if (name == "libretranslate.apikey")
    {
      if (!mLibreTranslateApiKey.empty())
      {
        std::cerr << "Error: API key for LibreTranslate is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      mLibreTranslateApiKey = value;
    } // if libretranslate.apikey
    else if (name == "giphy.apikey")
    {
      if (!mGiphyApiKey.empty())
      {
        std::cerr << "Error: API key for Giphy is specified more than once in file "
                  << fileName << "!" << std::endl;
        return false;
      }
      mGiphyApiKey = value;
    } // if giphy.apikey
    else
    {
      std::cerr << "Error while reading configuration file " << fileName
                << ": There is no setting named \"" << name << "\"!" << std::endl;
      return false;
    } // else (unrecognized setting name)
  } // while

  // Check presence of settings.
  if (mHomeServer.empty() || mUserId.empty() || mPassword.empty())
  {
    std::cerr << "Error: Some settings are missing in configuration file "
              << fileName << ", and thus botvinnik will not be able to work"
              << " properly." << std::endl;
    return false;
  } // if a setting is missing

  if (mStopUsers.empty())
  {
    std::cerr << "Error: bot.stop.allowed.userid has not been specified in "
              << "configuration file " << fileName << "!" << std::endl;
    return false;
  }

  if (mStopUsers.find(mUserId) == mStopUsers.end())
  {
    mStopUsers.insert(mUserId);
    std::clog << "Info: Adding user '" << mUserId << "' to list of users that"
              << " are allowed to stop the bot." << std::endl;
  }

  // Prefix may be missing. Set it to "!" in that case.
  if (mPrefix.empty())
  {
    mPrefix = "!";
    std::clog << "Info: Setting command prefix to '" << mPrefix
              << "', because none is given in the configuration file." << std::endl;
  }
  // Sync fail count may be missing. Set it to default in that case.
  if (mAllowedFailsIn64 < 0)
  {
    mAllowedFailsIn64 = default_allowed_failures;
    std::clog << "Info: Setting number of allowed sync failures to " << mAllowedFailsIn64
              << ", because none is given in the configuration file." << std::endl;
  }
  // Synchronization delay may be missing. Set it to the default in that case.
  if (mSyncDelay == std::chrono::milliseconds::zero())
  {
    mSyncDelay = default_sync_delay;
  }

  // Everything is good, so far.
  return true;
}

bool Configuration::load(const std::string& fileName)
{
  namespace fs = std::filesystem;

  std::string realName(fileName);
  findConfigurationFile(realName);
  if (realName.empty())
  {
    std::cerr << "Error: No configuration file was found!" << std::endl;
    return false;
  }
  try
  {
    fs::path p(realName);
    if (!fs::exists(p) || !fs::is_regular_file(p))
    {
      std::cerr << "Error: Configuration file " << realName
                << " does not exist or is not a regular file!" << std::endl;
      return false;
    }
  }
  catch (const fs::filesystem_error& ex)
  {
    std::cerr << "File system error while checking existence of file "
              << realName << ": " << ex.what() << std::endl;
  } // try-catch

  // clear any existing information
  clear();

  // load core configuration file
  return loadCoreConfiguration(realName);
}

void Configuration::clear()
{
  mHomeServer.clear();
  mUserId.clear();
  mPassword.clear();
  mStopUsers.clear();
  mAllowedFailsIn64 = -1;
  mSyncDelay = std::chrono::milliseconds::zero();
  mLibreTranslateServer.clear();
  mLibreTranslateApiKey.clear();
  mGiphyApiKey.clear();
}

} // namespace
